LFS-671: Script for importing cardiac rehab questionnaires

acrowthe opened the pull request:
Import cardiac rehab forms from .csv format
  - Detect form submission limits from title
  - Forms based on Report-type, sections based on sub-report
    - Sections with different submission limits than parent will become separate forms
Adjusted json_to_xml
  - Sanitize strings in more field locations
  - Improve detection of strings which need sanitizing
  - Add license to exported .xml file

**Testing**
Testing branch is [LFS-671-test](https://github.com/ccmbioinfo/lfs/tree/LFS-671-test), which includes LFS-731, LFS-733, LFS-734 and LFS-735.

To view only cardiac rehab questionnaires, use runmodes `nolfs` and `cardiac_rehab`: 
`java -jar distribution/target/lfs-*.jar -Dsling.run.modes=nolfs,cardiac_rehab`

**Known issues**
- Unnecessary "Other" questions due to other text field in option lists

acrowthe committed:
LFS-671: Script for importing cardiac rehab questionnaires
Import cardiac rehab forms from .csv format
- Detect form submission limits from title
- Forms based on Report-type, sections based on sub-report
  - Sections with different submission limits than parent will become seperate forms
Adjusted json_to_xml
- Sanitize strings in more field locations
- Improve detection of strings which need sanitizing
- Add license to exported .xml file

acrowthe committed:
LFS-671: Script for importing cardiac rehab questionnaires
Fix issue with "other" text field getting added to all lists

k-pace commented:
**Minor/picky thing**
- change text for "If only year is known, code as 01/01/YYYY" to "If only year is known, code as YYYY/01/01" to be consistent with date picker format (unless this is a local thing and different on every browser)
![image](https://user-images.githubusercontent.com/33154316/100279299-42c5db80-2f34-11eb-8dc4-fe461bc763d0.png)
**Bug?**
- Adverse Events forms are displaying as incomplete when all questions have a response
- to reproduce: create an adverse event form. Complete all fields. Save. Return to user dashboard. The form is under 'Incomplete Forms'

acrowthe commented:
> * change text for "If only year is known, code as 01/01/YYYY" to "If only year is known, code as YYYY/01/01" to be consistent with date picker format (unless this is a local thing and different on every browser)

This is a locale issue, but I'll at least adjust the text to read "If only year is known, code the month and day as 01"

acrowthe committed:
LFS-671: Script for importing cardiac rehab questionnaires
Re-import data
- Strip extra whitespace from option values and text
- Added study visit field to multiple forms
- Fixed typos and missing event multiple entry capabilities
- Fix empty edescription fields

sashaandjic commented:
DOB on Participant status is missing calendar gadget
![image](https://user-images.githubusercontent.com/23243244/100399451-e12f6b00-3020-11eb-97bc-9f42662bbb85.png)

sashaandjic commented:
Events data form: Medication change and Clinical Event sections are empty
![image](https://user-images.githubusercontent.com/23243244/100399739-eb059e00-3021-11eb-9e64-651357ab67fa.png)

sashaandjic commented:
Selecting N**one of the above** should deselect all other options
![image](https://user-images.githubusercontent.com/23243244/100400207-b1ce2d80-3023-11eb-976d-4c7ea1dfa736.png)

sashaandjic commented:
This could be problematic. Can we add separate exercise device setting as per Exercise device selection?
![image](https://user-images.githubusercontent.com/23243244/100400581-06be7380-3025-11eb-840f-8ffefc625741.png)

sashaandjic commented:
cpet_date missing
![image](https://user-images.githubusercontent.com/23243244/100400920-f955b900-3025-11eb-8196-d6a7993961a0.png)
![image](https://user-images.githubusercontent.com/23243244/100400951-112d3d00-3026-11eb-8469-5a3a0f898ea9.png)

marta- commented:
> Participant status data form:
> 
> * DOB missing calendar gadget
> * height, weight and waist: prevent entering values < 0; add units of measurements?

DOB is in month and year only, for which we don't have calendars. This should be specified in the description, otherwise it's indeed confusing.

Agreed re unit of measurement for wight and height.

marta- commented:
> Events data form: Medication change and Clinical Event sections are empty

I guess the bug here is that the section title is displayed for hidden conditional sections. Once you select the type of event in the first question, the corresponding section is correctly populated and displayed.

Created https://phenotips.atlassian.net/browse/LFS-746.

marta- commented:
> This could be problematic. Can we add separate exercise device setting as per Exercise device selection?

I agree this would be nice to have, but it complicates things too much. This is how the client designed the form and the information needs to end up in the same variable regardless of equipment type. We should leave it as it is.

marta- commented:
> cpet_date missing

It is not missing, it is displayed conditionally when the user answers Yes to the question "Has the patient had a previous (non-study) CPET"

marta- commented:
> Selecting N**one of the above** should deselect all other options

We don't have support for this yet, but I expect it to come up in other projects as well. Created am issue for it: https://phenotips.atlassian.net/browse/LFS-747 . Not sure if we'll implement it in this iteration though.

sashaandjic commented:
This bug is observed in this branch only. Steps to reproduce:
- create patient 
- create Participant Status form for the patient
- now try to create Participant Status form for the same patient
- an error msg appears **correctly** "Patient P1 already has 1 Participant Status form(s) filled out."
- click on the BACK bttn
- now try to select another form
- nothing happens - **wrong**
- click on CANCEL bttn
- click on + bttn to create a new form
- dialog opens up but it is unresponsive - **wrong**

marta- commented:
> This bug is observed in this branch only. Steps to reproduce:

Did you check the LFS-731 branch? Sounds like an issue with the introduction of material table for questionnaires.

sashaandjic commented:
@marta- 
Yes the bug is in LFS-731 branch.

acrowthe committed:
LFS-671: Script for importing cardiac rehab questionnaires
Reimport Participant status form to add units and DoB description

acrowthe committed:
LFS-671: Script for importing cardiac rehab questionnaires
Add min and max values

sashaandjic commented:
Baseline Health Information: Medications data form
- there are 3 "Name" fields that should have better labels ( Beta blocker name, Other med name, ? Other med name)
- also the first Name (beta-blocker_name) is required if beta blocker were checked in med types
- deselecting beta blockers should empty the corresponding fields (name, dose, unit,frequency, patient not taking med)
![image](https://user-images.githubusercontent.com/23243244/100482027-15a83300-30c4-11eb-9015-7471ed1e2301.png)

acrowthe committed:
LFS-671: Script for importing cardiac rehab questionnaires
Adjust Baseline Health Information - Medication question labels and names
Convert min and max values from strings to doubles
Adjust surveys mrb6 question from two 0-valued options to a single 0 valued option

acrowthe commented:
> * there are 3 "Name" fields that should have better labels ( Beta blocker name, Other med name, ? Other med name)

Changed to `Beta-blocker name`, and `Medication 1/2 name`

> * also for the the first Name (beta-blocker_name) fields are required if beta blocker were checked in med types above
> * deselecting beta blockers should empty the corresponding fields (name, dose, unit,frequency, patient not taking med)

Tested, these fields do seem to be being cleared when Beta-blocker is deselected

Also addressed a few items Marta mentioned. Those are listed in the most recent commit message

sashaandjic commented:
These 3 computed fields are blank even though I entered valid options for their computation
![image](https://user-images.githubusercontent.com/23243244/100491353-d17f5780-30f0-11eb-93e9-3a6360555630.png)

sashaandjic commented:
In the spreadsheet there are only 2 options: Yes and No
![image](https://user-images.githubusercontent.com/23243244/100492545-45bef880-30fb-11eb-94e3-6ea360ee90e6.png)
Also med reason field is not present
![image](https://user-images.githubusercontent.com/23243244/100492577-89b1fd80-30fb-11eb-978a-3c5bd7c10485.png)

acrowthe commented:
> In the spreadsheet there are only 2 options: Yes and No

All those options are present in the spreadsheet: The cell just isn't big enough to display all of them
![image](https://user-images.githubusercontent.com/20056751/100492641-26749b00-30fc-11eb-9293-0b8d870cb5fb.png)

acrowthe commented:
> These dasi_computed allows only one decimal number. As soon as there are to "Yes" options with the decimal score, the field is not computed.

Cannot reproduce. So long as all questions between `DASI date` and `DASI Computed` are filled out, `DASI computed` has a value from my testing. All fields being yes works:
![image](https://user-images.githubusercontent.com/20056751/100616566-44eaba00-32e7-11eb-926a-ba17677368e3.png)

Did notice a different issue though:
1. Fill out all the DASI questions with `No`. DASI computed will be 0 (correct), VO2peak and METs will not be filled out (incorrect)
2. Select `Yes` in one question. Computed, VO2Peak and METS will updated correctly
3. Replace `Yes` with `No`. Computed will reset to 0 (correct), VO2 and METs will not update (incorrect)

sashaandjic commented:
@acrowthe To reproduce the problem select Yes in the first 2 DASI questions.

acrowthe commented:
> Did notice a different issue though:
> 
> 1. Fill out all the DASI questions with `No`. DASI computed will be 0 (correct), VO2peak and METs will not be filled out (incorrect)
> 2. Select `Yes` in one question. Computed, VO2Peak and METS will updated correctly
> 3. Replace `Yes` with `No`. Computed will reset to 0 (correct), VO2 and METs will not update (incorrect)

Created LFS-751 to resolve this issue

sashaandjic commented:
Should visits 2,3,4 and 7 be invalid entries?
![image](https://user-images.githubusercontent.com/23243244/100691829-887d0c80-3357-11eb-9aae-0507f21effca.png)

sashaandjic commented:
Missing option Allergies:
![image](https://user-images.githubusercontent.com/23243244/100692833-aba8bb80-3359-11eb-85bb-8614ac646787.png)

sashaandjic commented:
Unlimited or 8?
![image](https://user-images.githubusercontent.com/23243244/100695162-a732d180-335e-11eb-844e-445d67862ed1.png)

marta- commented:
> Should visits 2,3,4 and 7 be invalid entries?

I would say there's no need for that.

k-pace commented:
> > Should visits 2,3,4 and 7 be invalid entries?
> 
> 
> 
> I would say there's no need for that.

Agreed - overkill. The only data entry users are trusted and have designed the study themselves

marta- commented:
> Unlimited or 8?

Unlimited sounds right, however visit number should not be there.

acrowthe committed:
LFS-671: Script for importing cardiac rehab questionnaires
Re-import forms for text changes and removing extra study visit questions

sashaandjic reviewed (approved):
tested; all good

one small suggestion for the 2 Baseline Health Info forms: Medications and Card params
![image](https://user-images.githubusercontent.com/23243244/100789023-e22b1880-33e3-11eb-9b8b-b81603366d3d.png)
![image](https://user-images.githubusercontent.com/23243244/100789911-16eb9f80-33e5-11eb-8be9-0ab639015868.png)
